fix: `trailingSlash` and child routes not working

pi0 opened the pull request:
<!--- Provide a general summary of your changes in the title above -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (a non-breaking change which fixes an issue)
- [ ] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)


## Description
Resolves #8422 also adding dev test for 3 possible values of trailing slash

`[vue-router] Named Route 'posts' has a default child route. When navigating to this named route (:to="{name: 'posts'"), the default child route will not be rendered. Remove the name from this route and use the name of the default child route for named links instead.`


## Checklist:
<!--- Put an `x` in all the boxes that apply. -->
<!--- If your change requires a documentation PR, please link it appropriately -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [ ] I have added tests to cover my changes (if not applicable, please state why)
- [ ] All new and existing tests are passing.

pi0 committed:
test: add failing test for trailing-slash and children

pi0 committed:
fix(utils): correctly detect default childRoute

pi0 committed:
fix typo in test

rchl commented:
Maybe you could also add trailingSlash=false version of that testsuite? I guess we don't have any existing testing for it?

pi0 commented:
@rchl I believe there is a test suit for children without trailing slash:
- https://github.com/nuxt/nuxt.js/tree/dev/test/fixtures/children
- https://github.com/nuxt/nuxt.js/blob/dev/test/dev/children.test.js

rchl commented:
`trailingSlash=undefined` and `trailingSlash=false` are not handled the same though.

farnabaz commented on packages/utils/src/route.js:
```diff
@@ -80,15 +80,14 @@ function cleanChildrenRoutes (routes, isChild = false, routeNameSplitter = '-',
+      const defaultChildRoute = route.children.find(child => child.path === '/' || child.path === '')
       const routeName = route.name
       if (defaultChildRoute) {
         if (trailingSlash === false) {
```
I think we can remove this condition now. We are going to remove `route.name` therefore we can use its name for the default child

pi0 committed:
test: update route.test

pi0 committed:
chore: remove extra condition

pi0 commented:
@rchl There is also a unit/snapshot test for routes utility: https://github.com/nuxt/nuxt.js/blob/dev/packages/utils/test/route.test.js

pi0 commented on packages/utils/test/route.test.js:
```diff
@@ -210,7 +210,6 @@ describe('util: route', () => {
       'pages/subpage/_param.vue',
       'pages/snake_case_route.vue',
       'pages/another_route/_id.vue',
-      'pages/another_route/_id.vue',
```
@clarkdo I think duplicate line was typo right?

rchl commented:
> @rchl There is also a unit/snapshot test for routes utility: [https://github.com/nuxt/nuxt.js/blob/dev/packages/utils/test/route.test.js](https://github.com/nuxt/nuxt.js/blob/dev/packages/utils/test/route.test.js?rgh-link-date=2020-12-01T16%3A59%3A41Z)

I don't see it testing the `trailingSlash: false` case which, as I've said, is not entirely identical to it being undefined.

pi0 committed:
test: add trailing-slash-false fixture

pi0 commented:
Actually [it is](https://github.com/nuxt/nuxt.js/blob/34359811f67a1ad62d4f705475652e26aa44a2d9/packages/utils/test/route.test.js#L237) :D Anyway agreed that snapshot couldn't cover this issue so added `trailing-slash-false` dev tests :)

pi0 commented on test/dev/trailing-slash.test.js:
```diff
@@ -0,0 +1,88 @@
+
+  test('/posts', async () => {
+    const { html } = await nuxt.server.renderRoute('/posts')
+    expect(html).toContain('[pages/posts]')
```
This is kinda wired behavior tough...

pi0 committed:
test: multi-variant trailing-slash test

danielroe commented:
I think we should pair this PR with another one for the docs to explain what the behaviour should be as I think there are a variety of different expectations in the related issues...

Happy to pick that up.

danielroe commented on test/fixtures/trailing-slash/trailish-slash.test.js:
```diff
@@ -0,0 +1,5 @@
+import { buildFixture } from '../../utils/build'
```
should be trailing-slash.test.js (?)

pi0 replied:
Nice fine! trailish is seems likely name for current status 😆

pi0 committed:
Merge branch '2.x' into fix/child-trailing

pi0 committed:
Merge branch 'dev' into fix/child-trailing

pi0 committed:
chore: fix typo in file

pi0 committed:
update test

modification scenario is wrong. not only because rootDir can contain other projects that build parallel, but also modules may generate a file outside

pi0 committed:
test: update windows snapshot
